### ISSUE TITLE
Waybar exits when returning from sleep

### ISSUE THREAD

**Issue by rnestler** [Alexays/Waybar#2310]
I have an issue where waybar stops (without segfault) when returning from sleep. Maybe also the removal from the dock and thus screens is involved. The output of waybar is the following:

```
# before suspend
[2023-07-13 09:03:45.201] [info] Bar configured (width: 2560, height: 27) for output: DP-3
[2023-07-13 09:03:45.202] [info] Bar configured (width: 2560, height: 27) for output: DP-4
[2023-07-13 09:03:45.202] [info] Bar configured (width: 1920, height: 27) for output: eDP-1

(waybar:168515): Gtk-CRITICAL **: 09:12:01.031: gtk_widget_set_accel_path: assertion 'GTK_IS_ACCEL_GROUP (accel_group)' failed
# a lot of these in between
(waybar:168515): Gtk-CRITICAL **: 09:48:51.292: gtk_widget_add_accelerator: assertion 'GTK_IS_ACCEL_GROUP (accel_group)' failed

# probably still before suspend
warning: queue 0x7f3228000ca0 destroyed while proxies still attached:
  zwp_primary_selection_offer_v1@4278190081 still attached
  wl_data_offer@4278190080 still attached
  zwp_primary_selection_device_v1@30 still attached
  zwp_primary_selection_device_manager_v1@22 still attached
  wl_data_device@29 still attached
  wl_seat@27 still attached
  wl_data_device_manager@25 still attached
  wl_registry@21 still attached
warning: queue 0x560c95272e30 destroyed while proxies still attached:
  wl_buffer@41 still attached
  wl_shm_pool@42 still attached
  xdg_wm_base@33 still attached
  wl_output@16 still attached
  wl_output@15 still attached
  wl_output@14 still attached
  wl_seat@13 still attached
  xdg_activation_v1@12 still attached
  wp_viewporter@11 still attached
  zwp_text_input_manager_v3@10 still attached
  zwp_pointer_constraints_v1@9 still attached
  zwp_relative_pointer_manager_v1@8 still attached
  zxdg_decoration_manager_v1@7 still attached
  wl_subcompositor@6 still attached
  wl_compositor@5 still attached
  wl_shm@4 still attached
  wl_registry@2 still attached
warning: queue 0x560c9527a520 destroyed while proxies still attached:
  wl_callback@45 still attached
[2023-07-13 10:54:48.566] [info] Bar removed from output: DP-3
[2023-07-13 10:54:48.569] [info] Bar removed from output: DP-4

# probably after resume

warning: queue 0x7f1198000ca0 destroyed while proxies still attached:
  zwp_primary_selection_offer_v1@4278190081 still attached
  wl_data_offer@4278190080 still attached
  zwp_primary_selection_device_v1@24 still attached
  zwp_primary_selection_device_manager_v1@20 still attached
  wl_data_device@23 still attached
  wl_seat@22 still attached
  wl_data_device_manager@21 still attached
  wl_registry@19 still attached
warning: queue 0x5556c165ae30 destroyed while proxies still attached:
  wl_buffer@42 still attached
  wl_shm_pool@39 still attached
  xdg_wm_base@31 still attached
  wl_output@14 still attached
  wl_seat@13 still attached
  xdg_activation_v1@12 still attached
  wp_viewporter@11 still attached
  zwp_text_input_manager_v3@10 still attached
  zwp_pointer_constraints_v1@9 still attached
  zwp_relative_pointer_manager_v1@8 still attached
  zxdg_decoration_manager_v1@7 still attached
  wl_subcompositor@6 still attached
  wl_compositor@5 still attached
  wl_shm@4 still attached
  wl_registry@2 still attached
warning: queue 0x5556c168da60 destroyed while proxies still attached:
  wl_callback@40 still attached
warning: queue 0x7f157c000ca0 destroyed while proxies still attached:
  zwp_primary_selection_offer_v1@4278190081 still attached
  wl_data_offer@4278190080 still attached
  zwp_primary_selection_device_v1@24 still attached
  zwp_primary_selection_device_manager_v1@20 still attached
  wl_data_device@23 still attached
  wl_seat@22 still attached
  wl_data_device_manager@21 still attached
  wl_registry@19 still attached
warning: queue 0x5644f1e41e30 destroyed while proxies still attached:
  wl_buffer@42 still attached
  wl_shm_pool@39 still attached
  xdg_wm_base@31 still attached
  wl_output@14 still attached
  wl_seat@13 still attached
  xdg_activation_v1@12 still attached
  wp_viewporter@11 still attached
  zwp_text_input_manager_v3@10 still attached
  zwp_pointer_constraints_v1@9 still attached
  zwp_relative_pointer_manager_v1@8 still attached
  zxdg_decoration_manager_v1@7 still attached
  wl_subcompositor@6 still attached
  wl_compositor@5 still attached
  wl_shm@4 still attached
  wl_registry@2 still attached
warning: queue 0x5644f1e74a60 destroyed while proxies still attached:
  wl_callback@40 still attached
warning: queue 0x7f0ffc000ca0 destroyed while proxies still attached:
  zwp_primary_selection_offer_v1@4278190083 still attached
  wl_data_offer@4278190082 still attached
  zwp_primary_selection_device_v1@28 still attached
  zwp_primary_selection_device_manager_v1@20 still attached
  wl_data_device@27 still attached
  wl_seat@26 still attached
  wl_data_device_manager@24 still attached
  wl_registry@19 still attached
warning: queue 0x55ea78fffe30 destroyed while proxies still attached:
  wl_buffer@41 still attached
  wl_shm_pool@39 still attached
  xdg_wm_base@31 still attached
  wl_output@14 still attached
  wl_seat@13 still attached
  xdg_activation_v1@12 still attached
  wp_viewporter@11 still attached
  zwp_text_input_manager_v3@10 still attached
  zwp_pointer_constraints_v1@9 still attached
  zwp_relative_pointer_manager_v1@8 still attached
  zxdg_decoration_manager_v1@7 still attached
  wl_subcompositor@6 still attached
  wl_compositor@5 still attached
  wl_shm@4 still attached
  wl_registry@2 still attached
warning: queue 0x55ea79032a60 destroyed while proxies still attached:
  wl_callback@43 still attached

# definitely after resume
(waybar:168515): LIBDBUSMENU-GLIB-WARNING **: 14:13:11.963: Unable to replace properties on 0: Error getting properties for ID

(waybar:168515): LIBDBUSMENU-GLIB-WARNING **: 14:13:11.965: Unable to replace properties on 0: Error getting properties for ID

(waybar:168515): LIBDBUSMENU-GLIB-WARNING **: 14:13:11.966: Unable to replace properties on 0: Error getting properties for ID

(waybar:168515): LIBDBUSMENU-GLIB-WARNING **: 14:13:12.066: Unable to replace properties on 0: Error getting properties for ID

(waybar:168515): LIBDBUSMENU-GLIB-WARNING **: 14:13:12.067: Unable to replace properties on 0: Error getting properties for ID

(waybar:168515): LIBDBUSMENU-GLIB-WARNING **: 14:13:12.069: Unable to replace properties on 0: Error getting properties for ID
[2023-07-13 14:13:12.072] [error] Invalid argument
```

**Comment by hurngchunlee** [comment]
Hi, I also run into the same issue recently after upgrading waybar to 0.9.19-1 on ArchLinux.  

By restarting waybar with log trace, I captured the following error.

At the time the laptop goes to sleep:

```
[2023-07-13 22:17:31.783] [debug] Output removed: Sharp Corporation 0x1516
[2023-07-13 22:17:31.784] [info] Bar configured (width: 1920, height: 37) for output: eDP-1
[2023-07-13 22:17:31.792] [info] Bar removed from output: eDP-1
```

followed by a repeating message like the one below:

```
(waybar:36830): Gdk-CRITICAL **: 22:17:31.820: gdk_monitor_get_scale_factor: assertion 'GDK_IS_MONITOR (monitor)' failed
```

and then

```
Fatal glibc error: pthread_mutex_lock.c:94 (___pthread_mutex_lock): assertion failed: mutex->__data.__owner == 0
[1]    36830 IOT instruction (core dumped)  waybar -l trace
```

Although the error is different, I also encountered few times the same error (i.e. `[error] Invalid argument` as posted by @rnestler).

**Comment by mysfitt** [comment]
I'm also seeing the same behavior after upgrading to 0.9.19-1 on Arch. I dug into the core dump and traced the problem to `waybar::util::SleeperThread::sleep_for` which appears to have changed in commit 3c9cbc9.

Please find my core file attached. I can easily reproduce the bug by going to suspend (S3) and then resuming. I'm on kernel 6.1.38-1-lts. Please let me know if I can help with debugging.
[core.waybar.zip](https://github.com/Alexays/Waybar/files/12053432/core.waybar.zip)

**Comment by mysfitt** [comment]
@rnestler I have confirmed that rolling back to version 0.9.18 resolves the issue. Until an actual fix is available, I'm going to keep locked on that version.

**Comment by Alexays** [comment]
@xytovl Maybe you can check, seem related to your work

**Comment by xytovl** [comment]
Do you know if this happens with a specific module or hardware configuration?
If you can post your waybar config I can try to reproduce on my machine. Also for the core file, I would need to know with which package this cas generated to get the associated debug symbols (distrib name + version should be enough).

**Comment by manfred3000** [comment]
@xytovl You've been pinged [here](https://github.com/Alexays/Waybar/issues/2304), as well. They also provided a core file using a minimal config...

**Comment by xytovl** [comment]
Issues are certainly from the same root cause, can you check if https://github.com/Alexays/Waybar/pull/2324 fixes this one too?

**Comment by ainola** [comment]
Arch maintainer here. Sorry for pushing that, all. I had tested but didn't think to test suspend/resume. I've included the destructor patch that seems to fix this (waybar-0.9.19-2)

Thanks for the patch, @xytovl

**Comment by mysfitt** [comment]
@Alexays Late to the party, but just wanted to confirm the fix worked for me as well. Thanks  to @xytovl for the quick fix and to @ainola for the quick patch to the Arch package. You guys are rock stars.
